Wikipedia/Code/AppDelegate.m

```diff
@@ -52,15 +51,52 @@ - (BOOL)application:(UIApplication*)application didFinishLaunchingWithOptions:(N
     [vc launchAppInWindow:self.window];
     self.appViewController = vc;
 
+    [self updateDynamicIconShortcutItems];
+    
     NSLog(@"%@", [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject]);
 
     return YES;
 }
 
+- (void)updateDynamicIconShortcutItems {
+    self.shortcutItemSelectedAtLaunch = nil;
+
+    if (![[UIApplication sharedApplication] respondsToSelector:@selector(shortcutItems)]) return;
+    
+    UIApplicationShortcutItem* (^makeShortcut)(NSString*, NSString*, NSString*, NSString*) = ^(NSString* type, NSString* title, NSString* subtitle, NSString* icon) {
+        return [[UIApplicationShortcutItem alloc] initWithType:type
+                                                localizedTitle:MWLocalizedString(title, nil)
+                                             localizedSubtitle:subtitle
+                                                          icon:[UIApplicationShortcutIcon iconWithTemplateImageName:icon]
+                                                      userInfo:nil];
+    };
+    
+    NSMutableArray* shortcutItems =
+    [[NSMutableArray alloc] initWithObjects:
+     makeShortcut(WMFIconShortcutTypeRandom, @"icon-shortcut-random-title", @"", @"random-quick-action"),
+     makeShortcut(WMFIconShortcutTypeNearby, @"icon-shortcut-nearby-title", @"", @"nearby-quick-action"),
+     nil
+     ];
+    
+    MWKTitle* lastRead = [[NSUserDefaults standardUserDefaults] wmf_openArticleTitle];
+    if (lastRead) {
+        [shortcutItems addObject:makeShortcut(WMFIconShortcutTypeContinueReading, @"icon-shortcut-continue-reading-title", lastRead.text, @"home-continue-reading-mini")];
+    }
+    
+    [shortcutItems addObject:makeShortcut(WMFIconShortcutTypeSearch, @"icon-shortcut-search-title", @"", @"search")];
+    
+    [UIApplication sharedApplication].shortcutItems = shortcutItems;
+}
+
+-(void)application:(UIApplication *)application performActionForShortcutItem:(UIApplicationShortcutItem *)shortcutItem completionHandler:(void (^)(BOOL))completionHandler {
+    self.shortcutItemSelectedAtLaunch = shortcutItem;
+}
+
 - (void)applicationWillResignActive:(UIApplication*)application {
     // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
     // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
     [[NSUserDefaults standardUserDefaults] wmf_setAppResignActiveDate:[NSDate date]];
+    [self updateDynamicIconShortcutItems];
```

Comment:
move this to background, no need to do this when app resigns active (i.e. phone call, notification, etc.)